### DISCUSSION TITLE
Tenant Extentions

### DISCUSSION THREAD

**Discussion by nsfuntionality** [JonPSmith/AuthPermissions.AspNetCore#18] [Q&A] [answered]
Hey Jon, Question regarding Tenant.

For Example, the tenant is only alow X number to invoices: does that mean I can add another property to the CompanyTenant model. or same that in a separate table with a data key? then how Admin will be able to change the X number once saved?

In another work, if I want to add more property on the CompanyTenant, which will be able to admin to edit and tenant to view.

**Comment by JonPSmith** [comment, marked as answer]
Hi @nsfuntionality,

Let me explain how AuthP works, which might help you with your questions.

The AuthP library contains the minimum data to handle Roles and Tenants linked to a AuthP User. From this you can a) control access to the features in your application via the `HadPermissions` attribute / method, and b) provides a `DataKey` to filter your application's data. If you want more features or information, then you can implement that in your application's DbContext. 

Another point is your application code and data shouldn't access the AuthP database - the only data you have are the Permissions and DataKey claims. I wanted to display the tenant name in the header of the  Example3 application, but I can't (shouldn't) access the AuthP data. That's why I added a `CompanyTenant` class in the Example3's application's DbContext, and I set it via the [InvoiceTenantChangeService](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/Example3.InvoiceCode/EfCoreCode/InvoiceTenantChangeService.cs). 

So to answer your questions:

- If you want to only allow X number of invoices, then you would store the X number in your application's DbContext using the DataKey so that the X is linked to a tenant (in Example3 I would put in in the `CompanyTenant`). You your code would read the X when the "Create Invoice" is clicked and check that the X limit hasn't been reached.
-  You would also build a service to allow an admin user to set / change the X number. You can use the AuthP's list of tenants, and add a "Set X" button next to each display

**Reply by nsfuntionality** [reply]
any example of the build service to allow an admin user to set / change

**Reply by JonPSmith** [reply]
OK, here are some options:

- You can use the AuthP's [`IAuthTenantAdminService`](https://github.com/JonPSmith/AuthPermissions.AspNetCore/blob/main/AuthPermissions/AdminCode/IAuthTenantAdminService.cs) service to list the tenants - see Example3's `TenantController` `Index` action. Then you can add a link / button (with the tenantId) to call to the action to set X.
- Your "Set X" method, which will be run by an admin user not linked to a tenant. Therefore you will have to use the `IgnoreQueryFilters` method to turn off the filtering and add its own filter (NOTE: The `Tenant` class has a method called `GetTenantDataKey`). This allows you to load the class with the X in (e.g. Example3's `CompanyTenant`) and update it.

**Reply by nsfuntionality** [reply]
Thank you, I think all i needed was to use the IgnoreQueryFilters method
